### ISSUE TITLE
Tips to reduce memory consumption in Auto-PyTorch

### ISSUE THREAD

**Issue by franchuterivera** [automl/Auto-PyTorch#259]
- [ ] We should not let the datamanager actively reside in memory when we are not using it. For example, there is no need to have a datamanager in [smbo](https://github.com/automl/Auto-PyTorch/blob/eda89f444611fd03767242ae0f12c294540e5f01/autoPyTorch/optimizer/smbo.py#L177).
- [ ] Also, after search has save the datamanager to disk, we can delete and garbage collect it.
- [ ] We also should datacollect and challenge the need of datamanager in the [evaluator](https://github.com/automl/Auto-PyTorch/blob/eda89f444611fd03767242ae0f12c294540e5f01/autoPyTorch/evaluation/tae.py#L171)
- [ ] We should improve the cross validation handling of the out of fold predictions. Rather than having a list that contains the OOF predicitons [here](https://github.com/automl/Auto-PyTorch/blob/eda89f444611fd03767242ae0f12c294540e5f01/autoPyTorch/evaluation/train_evaluator.py#L240) we should have a fixed array of n_samples created once at the beginning. OOF predictions from the k-fold model should be added smartly to this pre-existing array, something like `self.Y_optimization[test_indices] = opt_pred`. This ways predictions are sorted and can be used directly by ensemble selection without the need of saving this [array](https://github.com/automl/Auto-PyTorch/blob/eda89f444611fd03767242ae0f12c294540e5f01/autoPyTorch/evaluation/train_evaluator.py#L239)
- [ ] Calculating the train loss should be optional, not done by default [here](https://github.com/automl/Auto-PyTorch/blob/eda89f444611fd03767242ae0f12c294540e5f01/autoPyTorch/evaluation/train_evaluator.py#L257). We should prevent doing predict if not strictly needed.
- [ ] As reported already by @nabenabe0928 the biggest contribution comes from import files. In particular, just doing `import torch` consumes 2Gb of peak virtual memory and the majority of times this happens is for mypy typing. We should encapsulate these calls under `typing.TYPE_CHECKING` and only import the strictly needed class from pytorch.

**Comment by nabenabe0928** [comment]
Hi, thanks for raising this issue, I will move this issue to another memory consumption issue and close this issue once.
